File: CustomSteps/CreateJob/Java/src/com/esri/arcgis/jtx/customSteps/CreateJob.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package com.esri.arcgis.jtx.customSteps;

import com.esri.arcgis.geodatabase.esriRelCardinality;
import com.esri.arcgis.interop.AutomationException;
import com.esri.arcgis.interop.extn.ArcGISCategories;
import com.esri.arcgis.interop.extn.ArcGISExtension;
import com.esri.arcgis.jtx.IJTXActivityType;
import com.esri.arcgis.jtx.IJTXAuxProperties;
import com.esri.arcgis.jtx.IJTXAuxPropertiesProxy;
import com.esri.arcgis.jtx.IJTXAuxRecordContainer;
import com.esri.arcgis.jtx.IJTXConfigurationProperties;
import com.esri.arcgis.jtx.IJTXCustomStepFeedback;
import com.esri.arcgis.jtx.IJTXDatabase;
import com.esri.arcgis.jtx.IJTXCustomStepSafe;
import com.esri.arcgis.jtx.IJTXJob;
import com.esri.arcgis.jtx.IJTXJobManager;
import com.esri.arcgis.jtx.IJTXJobType;
import com.esri.arcgis.jtx.jtxAssignmentType;
import com.esri.arcgis.jtx.utilities.Constants;
import com.esri.arcgis.jtx.utilities.JTXUtilities;
import com.esri.arcgis.jtx.utilities.StepUtilities;
import com.esri.arcgis.jtx.utilities.WorkflowUtilities;
import java.io.IOException;

@ArcGISExtension(progid="JTXJavaSteps.CreateJob",
		clsid="77D1C6BD-4E47-42d5-9D39-31E40B5BE952",
		categories={ ArcGISCategories.JTXCustomStep })
public class CreateJob implements IJTXCustomStepSafe {
    private final String[] m_expectedArgs = {"jobtypeid", "assigngroup", "assignuser"};
    private IJTXDatabase m_ipDatabase = null;

    @Override
    public String getArgumentDescriptions() throws IOException, AutomationException {
        StringBuilder sb = new StringBuilder();
        sb.append("Job Type ID:\r\n");
        sb.append(String.format("\t/%s:<job type id> (required)\r\n\r\n", m_expectedArgs[0]));
        sb.append("Assign To Group\r\n");
        sb.append(String.format("\t/%s:<group to assign to> (optional)\r\n", m_expectedArgs[1]));
        sb.append("Assign To User:\r\n");
        sb.append(String.format("\t/%s:<username to assign to> (optional)\r\n", m_expectedArgs[2]));

        return sb.toString();
    }

    @Override
    public int execute(int jobID, int stepID, String[] argv, IJTXCustomStepFeedback is) throws IOException, AutomationException {
        String strValue = StepUtilities.getArgument(argv, m_expectedArgs[0], true);
        if (strValue == null) {
            //throw new ArgumentNullException(m_expectedArgs[0], String.Format("\nMissing the {0} parameter!", m_expectedArgs[0]));
            m_ipDatabase.logMessage(3, 1000, "Missing the {0} parameter!");
            return -1;
        }

        int jobTypeID;
        try {
            jobTypeID = Integer.parseInt(strValue);
        } catch (NumberFormatException ex) {
            m_ipDatabase.logMessage(3, 1000, ex.getMessage());
            return -1;
        }


        IJTXJobType pJobType = m_ipDatabase.getConfigurationManager().getJobTypeByID(jobTypeID);
        IJTXJobManager pJobMan = m_ipDatabase.getJobManager();
        IJTXJob pNewJob = pJobMan.createJob(pJobType, 0, true);

        IJTXActivityType pActType = m_ipDatabase.getConfigurationManager().getActivityType(Constants.ACTTYPE_CREATE_JOB);
        if (pActType != null) {
            pNewJob.logJobAction(pActType, null, "");
        }

        JTXUtilities.sendNotification(Constants.NOTIF_JOB_CREATED, m_ipDatabase, pNewJob, null);

        // Assign a status to the job if the Auto Assign Job Status setting is enabled
        IJTXConfigurationProperties pConfigProps = (IJTXConfigurationProperties) m_ipDatabase.getConfigurationManager();
        if (pConfigProps.propertyExists(Constants.JTX_PROPERTY_AUTO_STATUS_ASSIGN)) {
            String strAutoAssign = pConfigProps.getProperty(Constants.JTX_PROPERTY_AUTO_STATUS_ASSIGN);
            if (strAutoAssign.equals("TRUE")) {
                pNewJob.setStatus(m_ipDatabase.getConfigurationManager().getStatus("Created"));
            }
        }

        // Associate the current job with the new job with a parent-child relationship
        pNewJob.setParentJob(jobID);

        // Assign the job as specified in the arguments
        String strAssignTo = StepUtilities.getArgument(argv, m_expectedArgs[1], true);
        String strAssignToUser = StepUtilities.getArgument(argv, m_expectedArgs[2], true);
        if (strAssignTo != null)
        {
            m_ipDatabase.logMessage(5, 1000, "Assigning to group " + strAssignTo);
            pNewJob.setAssignedType(jtxAssignmentType.jtxAssignmentTypeGroup);
            pNewJob.setAssignedTo(strAssignTo);
        }
        else if (strAssignToUser != null)
        {
            m_ipDatabase.logMessage(5, 1000, "Assigning to user " + strAssignToUser);
            pNewJob.setAssignedType(jtxAssignmentType.jtxAssignmentTypeUser);
            pNewJob.setAssignedTo(strAssignToUser);
        }
        pNewJob.store();

        // Copy the workflow to the new job
        WorkflowUtilities.copyWorkflowXML(m_ipDatabase, pNewJob);

        // Create 1-1 extended property entries
        IJTXAuxProperties pAuxProps = new IJTXAuxPropertiesProxy(pNewJob);
        String[] contNames = pAuxProps.getContainerNames();
        for (String strContainerName : contNames) {
            IJTXAuxRecordContainer pAuxContainer = pAuxProps.getRecordContainer(strContainerName);
            if (pAuxContainer.getRelationshipType() == esriRelCardinality.esriRelCardinalityOneToOne) {
                pAuxContainer.createRecord();
            }
        }

        return 0;
    }

    @Override
    public String[] invokeEditor(int i, String[] inArgs) throws IOException, AutomationException {
        try {
            ArgEditor editorForm = new ArgEditor(null, m_ipDatabase, m_expectedArgs);
            String[] outArgs = editorForm.showDialog(inArgs);

            m_ipDatabase.logMessage(5, 1000, "There are " + outArgs.length + " arguments");
            return outArgs;
        } catch (IOException e) {
            if (m_ipDatabase != null) {
                m_ipDatabase.logMessage(3, 500, e.getMessage());
            }
            throw e;
        }
    }

    @Override
    public void onCreate(IJTXDatabase ipDatabase) throws IOException, AutomationException {
        m_ipDatabase = ipDatabase;
    }

    @Override
    public boolean validateArguments(String[] argv) throws IOException, AutomationException {
        String strValue = StepUtilities.getArgument(argv, m_expectedArgs[0], true);

        if (strValue == null) {
            return false;
        }
        return StepUtilities.areArgumentNamesValid(argv, m_expectedArgs);
    }

    @Override
    public String getName() throws IOException, AutomationException {
        return "JTXJavaSteps.CreateJob";
    }

    @Override
    public String getCLSID() throws IOException, AutomationException {
        return "{77D1C6BD-4E47-42d5-9D39-31E40B5BE952}";
    }
}

File: Classic/CustomSteps/CreateJob/CSharp/CreateJob.cs
/*Copyright 2015 Esri
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.?*/
using System;
using System.Text;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Windows.Forms;

using ESRI.ArcGIS.esriSystem;
using ESRI.ArcGIS.Geodatabase;

using ESRI.ArcGIS.JTX;
using ESRI.ArcGIS.JTX.Utilities;

namespace JTXSamples
{
    [Guid("9d4a3611-7455-485d-a583-3bfe5b66789b")]
    public class CreateJob : IJTXCustomStep
    {
        #region Registration Code
        [ComRegisterFunction()]
        static void Reg(String regKey)
        {
            ESRI.ArcGIS.JTX.Utilities.JTXUtilities.RegisterJTXCustomStep(regKey);
        }

        [ComUnregisterFunction()]
        static void Unreg(String regKey)
        {
            ESRI.ArcGIS.JTX.Utilities.JTXUtilities.UnregisterJTXCustomStep(regKey);
        }
        #endregion

        ////////////////////////////////////////////////////////////////////////
        // DECLARE: Data Members
        private readonly string[] m_expectedArgs = { "jobtypeid", "assigngroup", "assignuser" };
        private IJTXDatabase m_ipDatabase = null;

        #region IJTXCustomStep Members

        /// <summary>
        /// A description of the expected arguments for the step type.  This should
        /// include the syntax of the argument, whether or not it is required/optional, 
        /// and any return codes coming from the step type.
        /// </summary>
        public string ArgumentDescriptions
        {
            get
            {
                StringBuilder sb = new StringBuilder();
                sb.AppendLine(@"Job Type ID:");
                sb.AppendFormat("\t/{0}:<job type id> (required)\r\n\r\n", m_expectedArgs[0]);
                sb.AppendLine(@"Assign To Group:");
                sb.AppendFormat("\t/{0}:<group to assign to> (optional)\r\n", m_expectedArgs[1]);
                sb.AppendLine(@"Assign To User:");
                sb.AppendFormat("\t/{0}:<username to assign to> (optional)\r\n", m_expectedArgs[2]);

                return sb.ToString();
            }
        }

        /// <summary>
        /// Called when a step of this type is executed in the workflow.
        /// </summary>
        /// <param name="JobID">ID of the job being executed</param>
        /// <param name="StepID">ID of the step being executed</param>
        /// <param name="argv">Array of arguments passed into the step's execution</param>
        /// <param name="ipFeedback">Feedback object to return status messages and files</param>
        /// <returns>Return code of execution for workflow path traversal</returns>
        public int Execute(int JobID, int stepID, ref object[] argv, ref IJTXCustomStepFeedback ipFeedback)
        {
            System.Diagnostics.Debug.Assert(m_ipDatabase != null);

            string strValue = "";
            int jobTypeID = 0;

            if (!StepUtilities.GetArgument(ref argv, m_expectedArgs[0], true, out strValue))
            {
                throw new ArgumentNullException(m_expectedArgs[0], string.Format("\nMissing the {0} parameter!", m_expectedArgs[0]));
            }

            if (!Int32.TryParse(strValue, out jobTypeID))
            {
                throw new ArgumentNullException(m_expectedArgs[0], "Argument must be an integrer!");
            }

            IJTXJobType pJobType = m_ipDatabase.ConfigurationManager.GetJobTypeByID(jobTypeID);
            IJTXJobManager pJobMan = m_ipDatabase.JobManager;
            IJTXJob pNewJob = pJobMan.CreateJob(pJobType, 0, true);

            IJTXActivityType pActType = m_ipDatabase.ConfigurationManager.GetActivityType(Constants.ACTTYPE_CREATE_JOB);
            if (pActType != null)
            {
                pNewJob.LogJobAction(pActType, null, "");
            }

            JTXUtilities.SendNotification(Constants.NOTIF_JOB_CREATED, m_ipDatabase, pNewJob, null);

            // Assign a status to the job if the Auto Assign Job Status setting is enabled
            IJTXConfigurationProperties pConfigProps = (IJTXConfigurationProperties)m_ipDatabase.ConfigurationManager;
            if (pConfigProps.PropertyExists(Constants.JTX_PROPERTY_AUTO_STATUS_ASSIGN))
            {
                string strAutoAssign = pConfigProps.GetProperty(Constants.JTX_PROPERTY_AUTO_STATUS_ASSIGN);
                if (strAutoAssign == "TRUE")
                {
                    pNewJob.Status = m_ipDatabase.ConfigurationManager.GetStatus("Created");
                }
            }

            // Associate the current job with the new job with a parent-child relationship
            pNewJob.ParentJob = JobID;
            
            // Assign the job as specified in the arguments
            string strAssignTo = "";
            if (StepUtilities.GetArgument(ref argv, m_expectedArgs[1], true, out strAssignTo))
            {
                pNewJob.AssignedType = jtxAssignmentType.jtxAssignmentTypeGroup;
                pNewJob.AssignedTo = strAssignTo;
            }
            else if (StepUtilities.GetArgument(ref argv, m_expectedArgs[2], true, out strAssignTo))
            {
                pNewJob.AssignedType = jtxAssignmentType.jtxAssignmentTypeUser;
                pNewJob.AssignedTo = strAssignTo;
            }
            pNewJob.Store();

            // Copy the workflow to the new job
            WorkflowUtilities.CopyWorkflowXML(m_ipDatabase, pNewJob);

            // Create 1-1 extended property entries
            IJTXAuxProperties pAuxProps = (IJTXAuxProperties)pNewJob;
            System.Array contNames = pAuxProps.ContainerNames;
            IEnumerator contNamesEnum = contNames.GetEnumerator();
            contNamesEnum.Reset();

            while (contNamesEnum.MoveNext())
            {
                string strContainerName = (string)contNamesEnum.Current;
                IJTXAuxRecordContainer pAuxContainer = pAuxProps.GetRecordContainer(strContainerName);
                if (pAuxContainer.RelationshipType == esriRelCardinality.esriRelCardinalityOneToOne)
                {
                    pAuxContainer.CreateRecord();
                }
            }

            m_ipDatabase.LogMessage(5, 1000, System.Diagnostics.Process.GetCurrentProcess().MainWindowTitle);

            // Update Application message about the new job
            if (System.Diagnostics.Process.GetCurrentProcess().MainWindowTitle.Length > 0) //if its not running in server
            {
                MessageBox.Show("Created " + pJobType.Name + " Job " + pNewJob.ID, "Job Created", MessageBoxButtons.OK, MessageBoxIcon.Information);
            }

            return 0;
        }

        /// <summary>
        /// Invoke an editor tool for managing custom step arguments.  This is
        /// an optional feature of the custom step and may not be implemented.
        /// </summary>
        /// <param name="hWndParent">Handle to the parent application window</param>
        /// <param name="argsIn">Array of arguments already configured for this custom step</param>
        /// <returns>Returns a list of newely configured arguments as specified via the editor tool</returns>
        public object[] InvokeEditor(int hWndParent, object[] argsIn)
        {
            JTXSamples.ArgEditor editorForm = new JTXSamples.ArgEditor(m_ipDatabase, m_expectedArgs);
            object[] newArgs = null;

            return (editorForm.ShowDialog(argsIn, out newArgs) == DialogResult.OK) ? newArgs : argsIn;
        }

        /// <summary>
        /// Called when the step is instantiated in the workflow.
        /// </summary>
        /// <param name="ipDatabase">Database connection to the JTX repository.</param>
        public void OnCreate(IJTXDatabase ipDatabase)
        {
            m_ipDatabase = ipDatabase;
        }

        /// <summary>
        /// Method to validate the configured arguments for the step type.  The
        /// logic of this method depends on the implementation of the custom step
        /// but typically checks for proper argument names and syntax.
        /// </summary>
        /// <param name="argv">Array of arguments configured for the step type</param>
        /// <returns>Returns 'true' if arguments are valid, 'false' if otherwise</returns>
        public bool ValidateArguments(ref object[] argv)
        {
            string strValue = "";

            if (!StepUtilities.GetArgument(ref argv, m_expectedArgs[0], true, out strValue)) { return false; }
            return StepUtilities.AreArgumentNamesValid(ref argv, m_expectedArgs);
        }

        #endregion

    }	// End Class
}	// End Namespace
